pepper-apis/src/main/java/org/broadinstitute/ddp/export/collectors/ActivityResponseCollector.java

```diff
@@ -557,10 +566,14 @@ private void collectQuestionIntoRecord(Map<String, String> record, QuestionDef q
         }
     }
 
-    private void collectComponentIntoRecord(Map<String, String> record, ComponentBlockDef component, ComponentDataSupplier supplier) {
+    private void collectComponentIntoRecord(FormResponse formResponse, Map<String, String> record,
+                                            ComponentBlockDef component, ComponentDataSupplier supplier) {
         switch (component.getComponentType()) {
             case MAILING_ADDRESS:
-                record.putAll(addressFmt.collect(supplier.getAddress()));
+                MailAddress address = withAPIsTxn(handle -> {
```

Comment:
At this point in the code, we're in the middle of transforming the data to be exported, and we're very likely inside a transaction already, so opening another transaction is not good. We should query the data beforehand and pass it in. I think `ComponentDataSupplier` is the ideal place to store and pass along the activity snapshot data and the snapshotted address.

Reply:
I tried to check where this code called from. There are 3 methods (actually 4):
```
DataExporter.exportToElasticsearch()
   formatParticipantToStructuredJSON()
   formatParticipantToFlatJSON
DataExporter.exportDataSetAsCsv()
CustomExporter.exportDataSetAsCsv()
```
`ComponentDataSupplier` used in 3 of them except of `formatParticipantToStructuredJSON()`. Therefore I can not rely on this class 100%.
Another problem of calculating snapshot address before creating `ComponentDataSupplier` is that it needs to know `addressGuid` which corresponds to a certain `instanceId`. Therefore it needs to know activity instance data.
You advised to get from `Participant#activityInstanceSubstitutions` but we can use this substitution data only after it set to Participants. Possible way to do it: pass a callback to `Participant#putActivityInstanceSubstitutions` and in this callback to detect snapshotted address and save it to `ComponentDataSupplier` (but unfortunately not for a case of `formatParticipantToStructuredJSON()`)...

Reply:
Actually I did some refactoring: trying to get Participant#activityInstanceSubstitutions and detecting snapshotted address during creation of ComponentDataSupplier.